Dev to Main

invaderb opened the pull request:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: add endpoints to readme
chore: code clean up and refactor

nitzel committed:
chore[Readme]: make endpoints list more readable

nitzel committed:
fix: use path.join to prevent misconfigurations of e.g. DB_PATH when the ending slash is missing

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
feat: add rating endpoints
feat: add scheduled job
test: add tests for rating generation
chore: ad endpoints to readme

invaderb committed:
fix: comment typos

invaderb committed:
fix: comment typos

invaderb committed:
fix: select values
fix: add missing api param
fix: make optional columns

invaderb committed:
fix: select values
fix: add missing api param
fix: make optional columns

invaderb committed:
fix: swagger docs

invaderb committed:
fix: swagger docs

invaderb committed:
chore: remove unneeded console.log

invaderb committed:
chore: remove unneeded console.log

invaderb committed:
chore: code clean up and refactor
fix: add missing db names to typeorm feature and root

invaderb committed:
chore: code clean up and refactor
fix: add missing db names to typeorm feature and root

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
chore: code clean up

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
test: fix game service tests

invaderb committed:
test: fix game service tests

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
Merge branch 'rating-endpoints' of github.com:USTakAssociation/playtak-api into rating-endpoints

invaderb committed:
fix: issues from rebasing

invaderb committed:
chore: code clean up and refactor
chore: add better api property descriptions

invaderb committed:
fix: spelling error

invaderb committed:
fix: readme endpoints

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
feat: add bruno tests
fix: typos

invaderb committed:
Merge branch 'dev' into rating-endpoints

invaderb committed:
feat: add bru api test files

invaderb committed:
fix: fatigue rating calculation
chore: refactor to make variables more meaningful
chore: refactor ratings to have better typings
chore: refactor dto files
chore: update formatting
chore: fix typos

invaderb committed:
fix: unit test for updated typings
chore: fix typings
chore: clean up player dto

invaderb committed:
fix: typos
chore: update readme

invaderb committed:
chore: update readme

invaderb committed:
fix: logic for always adding fatiguerating to players who have or haven't changed
chore: add additional comments to explain the generateRatings

invaderb committed:
fix: add cron directive

invaderb committed:
fix: update deploy script

invaderb committed:
Merge pull request #8 from USTakAssociation/rating-endpoints

Feat: Player Rating Endpoints

invaderb committed:
chore: remove noisy debug log

invaderb committed:
fix: query param order

invaderb committed:
chore: rename fatiguerating to participation_rating
feat: rating list to  to json file

invaderb committed:
fix: set cron value from environment variable

invaderb committed:
chore: remove unneeded cron import

nitzel commented on .github/workflows/deploy-dev.yaml:
```diff
@@ -53,8 +56,12 @@ jobs:
             echo "ANON_DB_PATH = ${ANON_DB_PATH}" >> .env
             echo "DB_PATH = ${DB_PATH}" >> .env
             echo "APP_ENV = ${APP_ENV}" >> .env
+            echo "PREVIOUS_FILE = ${PREVIOUS_FILE}" >> .env
```
I'm surprised this works, I thought spaces around the `=` were not allowed or would lead to incorrect values being read. 🤷

invaderb replied:
🤷 I've never had issues adding the spaces but I can remove them  if there's a concern of an issue...

nitzel commented on previous.txt:
```diff
@@ -0,0 +1 @@
+547436 hash
```
should `previous.txt` be checked in? or added to `.gitignore`?

invaderb replied:
yeah probably should

nitzel commented on src/models/dto/events/events.dto.ts:
```diff
@@ -1,26 +1,36 @@
+/* eslint-disable indent */
```
maybe we should get rid of these nothing-saying `ApiProperty()`-decorators by letting swagger infer them for these basic types.

https://docs.nestjs.com/openapi/cli-plugin

Alternatively and/or additionally, we probably should try telling eslint to indent decorated functions correctly. This answer seems like a good start sometime in the future:
https://stackoverflow.com/a/72897089/5124673

invaderb replied:
oh I didn't know about this, I much prefer this option with the cli

nitzel replied:
don't fix it if it ain't broken 🤷‍♂️

nitzel commented on src/models/dto/players/player.dto.ts:
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable indent */
+import { ApiProperty } from "@nestjs/swagger";
+
+// Fatigue type with key of type string and value of type number
```
Type annotation already says the key/value types are `string/number`, but what do they represent? I believe adding that as a `/** */` comment would be more helpful.

invaderb replied:
Ah I accidentally left this in I'll update the comment to be more explicit

nitzel commented on src/models/dto/players/player.dto.ts:
```diff
@@ -0,0 +1,46 @@
+	participation_rating: number;
+		
+	@ApiProperty()
+	fatigue: Fatigue | string | null;
```
What does it mean if it's a string/null?

invaderb replied:
New registered players Fatigue is null until the rating script runs and adds it but when it gets parsed to json that where the Fatigue type comes in

nitzel replied:
Shouldn't this be `fatigue?: Fatigue` then?

invaderb replied:
@nitzel man it's a good thing I'm looking through this again from your comments I found a issue with the fatigue logic after adjusting this.

nitzel commented on src/models/ratings/ratings.service.spec.ts:
```diff
@@ -0,0 +1,352 @@
+				date: 1683328933,
+			}
+			const quickresult = {'R-0': 1, 'F-0': 1, '1-0': 1, '0-R': 0, '0-F': 0, '0-1': 0, '1/2-1/2': 0.5}[mockGame.result];
+			const player_white: Player = {
```
in the future we may consider refactoring this to

```ts
const get_default_player(id: number, overrides: Partial<Player>): Player => ({
    fatigue: {},
    boost: 0,
    ratedgames: 0,
    rating: 1000,
    maxrating: 1000,
    ratingage: 0,
    isbot: false,
    ratingbase: 1000,
    participation_rating: 1000,
    name: `Guest${id}`,
    id,
    ...overrides
});
const player_white = get_default_player(1, {
    ratingage: 1637395624533.48,
});
const player_black = get_default_player(2, {
    ratingage: 0,
    fatigue: {
        "1": 0.10901345963072347
    },
});
```

invaderb replied:
good suggestion it'll clean up a lot of duplicate code, Ill go ahead and implement this.

nitzel commented on src/models/ratings/tasks/rating.task.ts:
```diff
@@ -0,0 +1,41 @@
+	// | | hours
+	// | minutes
+	// seconds (optional)
+	//@Cron('0 15 * * * *')
```
why is this disabled?

nitzel commented on src/models/ratings/ratings.service.ts:
```diff
@@ -0,0 +1,492 @@
+			previous = await readFile(process.env.PREVIOUS_FILE, 'utf-8');
+		} catch (error) {
+			this.logger.error(error);
+			return;
```
shouldn't we raise the exception here for any downstream users instead of hiding it in the logs?

invaderb replied:
What do yo mean by downstream users?

invaderb replied:
@nitzel ah re-reading this I see what your referring to; the generate ratings currently isn't directly executed by a end user so throwing an actual error wouldn't do much, we are using new relic which will catch the logged error and can notify it, but it would also do the same for a thrown error so either would work in this.

I'll go ahead and add the throw error in case we do open up a endpoint to manually run the generate ratings

nitzel reviewed (commented):
Awesome, I like the `.bru` tests! 

Also crazy job you got the rating script running here! I'm still confused why fatigue can be string/Fatigue/null though ^^

It looks like the rating is neither running as a Cron nor does an endpoint exist to trigger it manually, so how are the ratings going to be calculated on production?

invaderb commented:
@nitzel fatigue gets saved as a string since its a json object but it can be null until a new player gets rated


Since the cron time gets pulled from the environment file you cant use the @cron directive its using the cron schedule registry  in the module init.

invaderb committed:
chore: add previous.txt to gitignore
chore: code clean up and refactor

invaderb committed:
fix: parsed fatigue
fix: update fatigue to return player
fix: player type class
fix: update fatigue unit test and types
test: add unit test for parsing the fatigue

invaderb committed:
chore: clean up
